romana/kubernetes/resources.go

```diff
@@ -0,0 +1,163 @@
+*/
+
+// Event is a representation of a structure that we receive from kubernetes API.
+type Event struct {
+	Type   string     `json:"Type"`
+	Object KubeObject `json:"object"`
+}
+
+const (
+	KubeEventAdded         = "ADDED"
+	KubeEventDeleted       = "DELETED"
+	InternalEventDeleteAll = "_DELETE_ALL"
+)
+
+// KubeObject is a representation of object in kubernetes.
+type KubeObject struct {
+	Kind       string            `json:"kind"`
+	Spec       Spec              `json:"spec"`
+	ApiVersion string            `json:"apiVersion"`
+	Metadata   Metadata          `json:"metadata"`
+	Status     map[string]string `json:"status,omitempty"`
+}
+
+// makeId makes id to identify kube object.
+func (o KubeObject) makeId() string {
+	id := o.Metadata.Name + "/" + o.Metadata.Namespace
+	return id
+}
+
+type PodSelector map[string]string
+
+type FromEntry struct {
+	Pods PodSelector `json:"pods"`
+}
+
+type AllowIncoming struct {
+	From    []FromEntry `json:"from"`
+	ToPorts []ToPort    `json:"toPorts"`
+}
+
+type ToPort struct {
+	Port     uint   `json:"port"`
+	Protocol string `json:"protocol"`
+}
+
+// TODO need to find a way to use different specs for different resources.
+type Spec struct {
+	AllowIncoming AllowIncoming `json:"allowIncoming"`
+	PodSelector   PodSelector   `json:"podSelector"`
+}
+
+// Metadata is a representation of metadata in kubernetes object
+type Metadata struct {
+	Name              string            `json:"name"`
+	Namespace         string            `json:"namespace"`
+	SelfLink          string            `json:"selfLink"`
+	Uid               string            `json:"uid"`
+	ResourceVersion   string            `json:"resourceVersion"`
+	CreationTimestamp string            `json:"creationTimestamp"`
+	Labels            map[string]string `json:"labels"`
+}
+
+// watchEvents maintains goroutine fired by NsWatch, restarts it in case HTTP GET times out.
+func (l *kubeListener) watchEvents(done <-chan Done, url string, resp *http.Response, out chan Event) {
+	log.Println("Received namespace related event from kubernetes", resp.Body)
+
+	dec := json.NewDecoder(resp.Body)
+	var e Event
+
+	for {
+		select {
+		case <-done:
+			return
+		default:
+			// Attempting to read event from HTTP connection
+			err := dec.Decode(&e)
+			if err != nil {
+				// If fail
+				log.Printf("Failed to decode message from connection %s due to %s\n. Attempting to re-establish", url, err)
+				// Then stop all goroutines
+				out <- Event{Type: InternalEventDeleteAll}
+
+				// And try to re-establish HTTP connection
+				resp, err2 := http.Get(url)
+				if err2 != nil {
+					log.Printf("Failed establish connection %s due to %s\n.", url, err)
+				} else if err2 == nil {
+					dec = json.NewDecoder(resp.Body)
+				}
```

Comment:
Need a sleep here. Otherwise, it will run through this loop few hundred times few hundred times between HTTP packet sent and response received, flooding logs with "Failed to establish connection ... " message